### PR TITLE
ft(two-factor-auth):generate 2fa to the user)

### PR DESCRIPTION
**Pull request by buka1calvin**
### What does this PR do?
- validating seller using two factor authentication :

1. seller after logging in should receive an extra validation code on the end point with token and a code(OTP)
2. Seller should use that OTP provided by email for verification

### Description of Task to be completed?

- [x] GIVEN there are multiple users on the system and a seller**

1. seller should be able to login:

- [x] using traditional ways and be able to generate a one time passowrd
- [x] user should be able to receive a email containing a code for validation
- here are some endpoints already working
- `/api/v1/users/login`:for generating a token and OTP when user Role:"seller"
- `/api/v1/users/login/validate/:token`:checking if the OTP provided is valide
### How should this be manually tested?

**Installation**
**1. Clone the repository by running** 
```
git clone https://github.com/atlp-rwanda/destructors-ec-bn.git
```
**2. Navigate to the project directory by running cd destructors-ec-bn.**
```
cd destructors-ec-bn
```
```
ft-two-factor-auth-184637138
```
**3. Install the necessary dependencies by running npm install.**
```
npm install
```
**4. Run the development server by running npm run dev. This will start the server using Nodemon, which will automatically reload the server when changes are made to the code.**
```
npm run dev
```
**5. Run the tests by running npm run test** 
```
npm run test
```
### Any background context you want to provide?
when running this on local server put into consideration the below information 👇 
add the token generated on login and pass it in the endpoint 
### What are the relevant pivotal tracker/Trello stories?
- Tracker: 184637138
### Screenshots (if appropriate)

- none

### Questions:

- none

### CODE REVIEW

File: src/services/emailValidation.service.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import sgMail from '@sendgrid/mail';
+import User from "../database/models/index.js"
+import {OTP} from "../database/models/index"
+
+const validOTPmail=async(user,otp,token)=>{
```
Comment (Abdulkeza) [inline src/services/emailValidation.service.js]: please let's use understandable function names and variable names

File: src/services/emailValidation.service.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      },
+      subject:'click here to confirm',
+      text:'this is the message from sendgrid',
+      html:`<a href="${url}">${url} </a>copy this code: ${otp}`
+    }
+    sgMail.send(message)
+    .then(res=>{
+      console.log("mesage sent ...")
+    })
+    .catch(error=>console.log(error))
```
Comment (Abdulkeza) [inline src/services/emailValidation.service.js]: we are supposed to have a function that sends emails, what you only need is to pass data and subject.  you can ask the team.

File: src/utils/verifyOTP.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+     const user=await OTP.findOne({
+        where:{
+            otp:otp,
+            createdAt:{
+                [Op.gte]:new Date(new Date()-10*60*1000),
+            },
+        },
+    })
+    if(!user){
+        return res.send("invalid OTP!")
```
Comment (Abdulkeza) [inline src/utils/verifyOTP.js]: please let's use the same structure  ie `res.status(statusCode).json({message: message})`

File: src/utils/verifyOTP.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { Op } from "sequelize";
+import {User,OTP} from "../database/models";
+const verifyOTP=async(req,res)=>{
+    const {email,otp}=req.body
+     const user=await OTP.findOne({
```
Comment (Abdulkeza) [inline src/utils/verifyOTP.js]: are you finding a user? or is OTP?

File: src/utils/verifyOTP.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        },
+    })
+    if(!user){
+        return res.send("invalid OTP!")
+    }
+    //clear the otp and log the user in
+    else{
+    user.otp=null;
+    await user.save()
+    res.send("logged in successfully")
```
Comment (Abdulkeza) [inline src/utils/verifyOTP.js]: Please you are supposed to send token once a user logged in! also remember to include status code and use JSON format

**Review by Abdulkeza — CHANGES_REQUESTED** [review]
left feedback bellow

File: src/services/emailValidation.service.js (hunk 2 above)
Comment (buka1calvin) [reply src/services/emailValidation.service.js]: i'm waiting for the team so that i can pull kairos changes he's the one to make that function

**Review by katros1 — APPROVED** [review]

File: src/services/emailValidation.service.js (hunk 2 above)
Comment (katros1) [reply src/services/emailValidation.service.js]: @buka1calvin try to check in service there a service for sending email

**Review by nkurunziza1 — APPROVED** [review]

**Review by ba-lambert — CHANGES_REQUESTED** [review]
Seller is the one have to receive the OTP not buyer

**Review by Abdulkeza — APPROVED** [review]

**Review by JENNY-257 — APPROVED** [review]

**Pull request merged**